File: 09_Tic_tac_toe.py
import random

# Display the board
def display_board(board):
    print('......................................')
    print(f'{board[0]} | {board[1]} | {board[2]}')
    print('---------')
    print(f'{board[3]} | {board[4]} | {board[5]}')
    print('---------')
    print(f'{board[6]} | {board[7]} | {board[8]}')

# Check if a player has won
def win(board, player):
    return win_line(board, player)

def win_line(board, player):
    # Rows
    if (board[0] == player and board[1] == player and board[2] == player) or \
       (board[3] == player and board[4] == player and board[5] == player) or \
       (board[6] == player and board[7] == player and board[8] == player):
        return True
    # Columns
    if (board[0] == player and board[3] == player and board[6] == player) or \
       (board[1] == player and board[4] == player and board[7] == player) or \
       (board[2] == player and board[5] == player and board[8] == player):
        return True
    # Diagonals
    if (board[0] == player and board[4] == player and board[8] == player) or \
       (board[2] == player and board[4] == player and board[6] == player):
        return True
    return False

# Check if the board is full
def full_board(board):
    return '-' not in board

# Get available moves
def available_moves(board):
    return [i + 1 for i, spot in enumerate(board) if spot == '-']

# Make a move
def make_move(board, move, player):
    board[move - 1] = player

# AI's turn (simple strategy: win, block, or random)
def ai_move(board, player):
    # Check if AI can win
    for move in available_moves(board):
        temp_board = board[:]
        make_move(temp_board, move, player)
        if win(temp_board, player):
            return move
    # Check if AI can block
    opponent = 'o' if player == 'x' else 'x'
    for move in available_moves(board):
        temp_board = board[:]
        make_move(temp_board, move, opponent)
        if win(temp_board, opponent):
            return move
    # Otherwise, choose a random available move
    return random.choice(available_moves(board))

# Game loop
def play_game(board, player):
    display_board(board)
    
    if win(board, 'x'):
        print('Player X wins!')
        return
    if win(board, 'o'):
        print('Player O wins!')
        return
    if full_board(board):
        print('It\'s a draw!')
        return
    
    if player == 'x':
        player_move(board)
        next_player = 'o'
    else:
        move = ai_move(board, 'o')
        make_move(board, move, 'o')
        next_player = 'x'
    
    play_game(board, next_player)

# Player's move
def player_move(board):
    while True:
        try:
            move = int(input('Enter your move (1-9): '))
            if move < 1 or move > 9:
                print('Invalid move. Try again.')
                continue
            if board[move - 1] != '-':
                print('This spot is already taken. Try again.')
                continue
            make_move(board, move, 'x')
            break
        except ValueError:
            print('Invalid input. Please enter a number between 1 and 9.')

# Start the game
def start_game():
    play_game(['-', '-', '-', '-', '-', '-', '-', '-', '-'], 'x')

# Start the game
start_game()


#OUTPUT:
# python filename.py

#Which Algo used
# Greedy Strategy: AI makes the best decision based on immediate benefits (win or block).
# Search: AI looks one step ahead (win or block).
# Recursion: The game flow is managed using recursion to alternate between players.

# Component 	    Algorithm	                           Key Feature
# Win Check	     Brute-force pattern matching    	Checks all 8 winning combinations
# AI Move	     Greedy heuristic                	Win-block-random priority
# Game Flow	     Recursive state management	        Alternates turns until termination
# Player Input	    Input validation	               Ensures legal moves

File: 09_Tic_tac_toe.pl
% Tic-Tac-Toe Game in Prolog
% Representation:
% Board: A list of 9 elements representing the board.
% '-' for empty, 'x' for player X, 'o' for player O.
% Moves: Numbers 1-9 representing the positions on the board.
% Display the board
display_board(Board) :-
nth0(0, Board, A), nth0(1, Board, B), nth0(2, Board, C),
nth0(3, Board, D), nth0(4, Board, E), nth0(5, Board, F),
nth0(6, Board, G), nth0(7, Board, H), nth0(8, Board, I),
format('......................................~n'),
format('~w | ~w | ~w~n', [A, B, C]),
format('---------~n'),
format('~w | ~w | ~w~n', [D, E, F]),
format('---------~n'),
format('~w | ~w | ~w~n', [G, H, I]).
% Check if a player has won
win(Board, Player) :-
win_line(Board, Player).
win_line(Board, Player) :-
% Rows
(nth0(0, Board, Player), nth0(1, Board, Player), nth0(2, Board, Player));
(nth0(3, Board, Player), nth0(4, Board, Player), nth0(5, Board, Player));
(nth0(6, Board, Player), nth0(7, Board, Player), nth0(8, Board, Player));
% Columns
(nth0(0, Board, Player), nth0(3, Board, Player), nth0(6, Board, Player));
(nth0(1, Board, Player), nth0(4, Board, Player), nth0(7, Board, Player));
(nth0(2, Board, Player), nth0(5, Board, Player), nth0(8, Board, Player));
% Diagonals
(nth0(0, Board, Player), nth0(4, Board, Player), nth0(8, Board, Player));
(nth0(2, Board, Player), nth0(4, Board, Player), nth0(6, Board, Player)).
% Check if the board is full
full_board(Board) :-
\+ member('-', Board).


% Get available moves
available_moves(Board, Moves) :-
findall(Move, (nth0(Index, Board, '-'), Move is Index + 1), Moves).
% Make a move
make_move(Board, Move, Player, NewBoard) :-
Index is Move - 1,
replace(Board, Index, Player, NewBoard).
% Replace an element in a list
replace(List, Index, Element, NewList) :-
replace(List, Index, Element, 0, NewList).
replace([_|Rest], 0, Element, _, [Element|Rest]).
replace([H|Rest], Index, Element, Count, [H|NewRest]) :-
NextCount is Count + 1,
NextIndex is Index - 1,
replace(Rest, NextIndex, Element, NextCount, NewRest).
% AI's turn (simple strategy: win, block, or random)
ai_move(Board, Move, Player) :-
% Check if AI can win
available_moves(Board, Moves),
member(Move, Moves),
make_move(Board, Move, Player, TempBoard),
win(TempBoard, Player), !.
ai_move(Board, Move, Player) :-
% Check if AI can block
opponent(Player, Opponent),
available_moves(Board, Moves),
member(Move, Moves),
make_move(Board, Move, Opponent, TempBoard),
win(TempBoard, Opponent), !.
ai_move(Board, Move, _) :-
% Otherwise, choose a random available move
available_moves(Board, Moves),
random_member(Move, Moves).

opponent(x, o).
opponent(o, x).
% Game loop
play_game(Board, Player) :-
display_board(Board),
(win(Board, x) -> format('Player X wins!~n');
win(Board, o) -> format('Player O wins!~n');
full_board(Board) -> format('It\'s a draw!~n');
(Player = x -> player_move(Board, NewBoard), NextPlayer = o;
Player = o -> ai_move(Board, Move, o), make_move(Board, Move, o, NewBoard),
NextPlayer = x),
play_game(NewBoard, NextPlayer)).
% Player's move
player_move(Board, NewBoard) :-
repeat,
write('Enter your move (1-9): '),
read(Move),
(integer(Move), Move >= 1, Move =< 9 ->
available_moves(Board, Moves),
member(Move, Moves),
make_move(Board, Move, x, NewBoard), !;
write('Invalid move. Try again.~n'), fail).
% Start the game
start_game :-
play_game(['-', '-', '-', '-', '-', '-', '-', '-', '-'], x).
%random member
random_member(X, L) :-
length(L, Len),
random(0, Len, Index),
nth0(Index, L, X).



# #OUTPUT
# 1) Start the game by running:
#     ?- start_game.




